bentoml/_internal/resource.py

```diff
@@ -0,0 +1,281 @@
+from __future__ import annotations
+
+import os
+import re
+import math
+import typing as t
+import logging
+import functools
+from abc import ABC
+from abc import abstractmethod
+
+import psutil
+
+from ..exceptions import BentoMLConfigException
+
+logger = logging.getLogger(__name__)
+
+_RESOURCE_REGISTRY: dict[str, t.Type[Resource[t.Any]]] = {}
+
+T = t.TypeVar("T")
+
+
+def get_resource(resources: dict[str, t.Any], resource_kind: str) -> t.Any:
+    if resource_kind not in _RESOURCE_REGISTRY:
+        raise BentoMLConfigException(f"Unknown resource kind '{resource_kind}'.")
+
+    resource: t.Type[Resource[t.Any]] = _RESOURCE_REGISTRY[resource_kind]
+
+    if resource_kind in resources:
+        if resources[resource_kind] == "system":
+            return resource.from_system()
+        else:
+            res = resource.from_spec(resources[resource_kind])
+            resource.validate(res)
+            return res
+    else:
+        return None
+
+
+def system_resources() -> dict[str, t.Any]:
+    res: dict[str, t.Any] = {}
+    for resource_kind, resource in _RESOURCE_REGISTRY.items():
+        res[resource_kind] = resource.from_system()
+    return res
+
+
+class Resource(t.Generic[T], ABC):
+    def __init_subclass__(cls, *, resource_id: str):  # pylint: disable=arguments-differ
+        _RESOURCE_REGISTRY[resource_id] = cls
+
+    @classmethod
+    @abstractmethod
+    def from_spec(cls, spec: t.Any) -> T:
+        """
+        Get an instance of this resource from user input. For example, a CPU resource
+        might parse "10m" and return a CPU resource with 0.01 CPUs.
+        """
+
+    @classmethod
+    @abstractmethod
+    def from_system(cls) -> T:
+        """
+        Infer resource value from the system.
+        """
+
+    @classmethod
+    @abstractmethod
+    def validate(cls, val: T):
+        """
+        Validate that the resources are available on the current system.
+        """
+
+
+class CpuResource(Resource[float], resource_id="cpu"):
+    @classmethod
+    def from_spec(cls, spec: t.Any) -> float:
+        """
+        Convert spec to CpuResource.
+
+        spec can be a float, int or string.
+        - 1.0 -> 1.0
+        - 1 -> 1.0
+        - "1" -> 1.0
+        - "10m" -> 0.01
+        """
+        if not isinstance(spec, (int, float, str)):
+            raise TypeError("cpu must be int, float or str")
+
+        if isinstance(spec, (int, float)):
+            return float(spec)
+
+        milli_match = re.match("([0-9]+)m", spec)
+        if milli_match:
+            return float(milli_match[1]) / 1000.0
+
+        try:
+            return float(spec)
+        except ValueError:
+            raise BentoMLConfigException(f"Invalid CPU resource limit '{spec}'. ")
+
+    @classmethod
+    @functools.lru_cache(maxsize=1)
+    def from_system(cls) -> float:
+        if psutil.POSIX:
+            return query_cgroup_cpu_count()
+        else:
+            return float(query_os_cpu_count())
```

Comment:
Pyright is complaining about this and I currently have no idea why.

Reply:
I think it's okay to remove lru_cache here?

Reply:
Yeah, I think that's fine. Just confused.

Reply:
I guess the lru_cache is not well type annotated to `_lru_cache_wrapper[float]`